src/async_fuse/memfs/kv_engine.rs

```diff
@@ -180,23 +248,229 @@ pub trait MetaTxn {
+        self.prev_kv = true;
+        self
+    }
+
+    #[allow(dead_code)]
+    #[must_use]
+    /// Set the range end of the delete operation
+    fn with_range_end(mut self, range_end: Vec<u8>) -> Self {
+        self.range_end = Some(range_end);
+        self
+    }
+
+    /// Delete all keys
+    #[allow(dead_code)]
+    #[must_use]
+    /// Set the range end of the delete operation
+    fn with_all_keys(mut self) -> Self {
+        self.range_end = Some(vec![0xff]);
+        self
+    }
+}
+
+/// `KeyRange` is an abstraction for describing etcd key of various types.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+#[allow(dead_code)]
+pub struct KeyRange {
+    /// The first key of the range and should be non-empty
+    pub(crate) key: Vec<u8>,
+    /// The key following the last key of the range
+    pub(crate) range_end: Vec<u8>,
+}
+
+impl KeyRange {
+    /// Creates a new `KeyRange` for describing a range of multiple keys.
+    #[inline]
+    #[allow(dead_code)]
+    pub fn range<K, R>(key: K, range_end: R) -> Self
+    where
+        K: Into<Vec<u8>>,
+        R: Into<Vec<u8>>,
+    {
+        Self {
+            key: key.into(),
+            range_end: range_end.into(),
+        }
+    }
+
+    /// Creates a new `KeyRange` for describing a specified key.
+    #[inline]
+    #[allow(dead_code)]
+    pub fn key<K>(key: K) -> Self
+    where
+        K: Into<Vec<u8>>,
+    {
+        Self {
+            key: key.into(),
+            range_end: vec![],
+        }
+    }
+
+    /// Creates a new `KeyRange` for describing all keys.
+    #[inline]
+    #[must_use]
+    #[allow(dead_code)]
+    pub fn all() -> Self {
+        Self {
+            key: vec![0],
+            range_end: vec![0],
+        }
+    }
+
+    /// Creates a new `KeyRange` for describing keys prefixed with specified value.
+    #[inline]
+    #[allow(dead_code)]
+    pub fn prefix<K>(prefix: K) -> Self
+    where
+        K: Into<Vec<u8>>,
+    {
+        let key = prefix.into();
+        if key.is_empty() {
+            // An empty Vec<u8> results in an invalid KeyRange.
+            // Assume that an empty value passed to this method implies no prefix (i.e., all keys).
+            return Self::all();
+        }
+
+        let mut first_value = true;
+        let mut range_end = key
+            .iter()
+            .rev()
+            .filter_map(|e| {
+                if *e < 0xff {
+                    if first_value {
+                        first_value = false;
+                        Some(e.overflow_add(1))
+                    } else {
+                        Some(*e)
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<u8>>();
+        range_end.reverse();
+        Self { key, range_end }
+    }
+
+    /// Take key value
+    #[inline]
+    #[allow(dead_code)]
+    pub(crate) fn take_key(&mut self) -> Vec<u8> {
+        std::mem::take(&mut self.key)
```

Comment:
Do you want to make sure the key is taken once? If your purpose is to avoid memory copy, it doens't work.